sapicore/engine/ensemble/__init__.py

```diff
@@ -0,0 +1,66 @@
+""" Ensembles (layers) manage neuron collections by fixing their size, extending attribute tensors,
+and handling layer-aware operations such as heterogeneous parameter initialization
+and normalization.
+
+Note
+----
+Ensemble instance attributes are 1D tensors corresponding to potentially heterogeneous parameters
+and states of neurons (tensor elements) whose behavior is prescribed by their parent class
+(e.g., :class:`~engine.neuron.spiking.LIF.LIFNeuron`).
+
+Ensembles can be initialized programmatically or from a configuration YAML containing neuron parameters and their
+generation method (fixed, zipped, grid, or drawn from a distribution). Forward calls invoke the parent neuron class
+method.
+
+Example
+-------
+Initialize an ensemble of basic analog neurons, inheriting default parameters and behavior from
+:class:`~engine.neuron.analog.AnalogNeuron`:
+
+    >>> from sapicore.engine.ensemble.spiking import LIFEnsemble
+    >>> layer = LIFEnsemble(num_units = 10)
+
+"""
+import torch
+from torch import Tensor
+
+from sapicore.engine.neuron import Neuron
+
+
+class Ensemble(Neuron):
+    """Ensemble base class. Provides generic implementations of tensor expansion and parameter diversification."""
+
+    def __init__(self, num_units: int = 1, **kwargs):
+        """Constructs a generic ensemble instance, inheriting attributes from :class:`~engine.neuron.Neuron`."""
+        super().__init__(**kwargs)
+        self.num_units = num_units
+
+        # expand configurable parameter tensors to size `num_units`.
+        for prop in self._config_props_:
```

Comment:
Same about using the functions to access these props (and elsewhere.)

Reply:
See above for why this doesn't work as a blanket recommendation. Changing to `_config_props` is fine here because the attributes have been initialized in the super call. Changing to `_loggable_props` will not, because of the mandatory-first-access-before-cache-update bug I found.

Reply:
Having realized that `loggable_props` should be used in this context (not `_loggable_props`--I think those were interchanged in one of your comments dealing with _config_props, leading me to wrestle with the _loggable cache instead), the fix has been applied throughout the code base.